### PR TITLE
Fix the running task counter for inlined functions with on-stmts

### PR DESCRIPTION
**Pull request by ronawho**
Previously we inserted a pair of running task modifiers for each call to
an outlined function. This was wrong for inlined functions where each
callsite of the function results in a call to the outlined function. Fix
that by tracking which outlined functions have already had their running
task counter modified. Before this change the following test would
incorrectly report 10 runningTasks, but now correctly reports 1.

```chpl
inline proc movedOn() {
  var count: int;
  on Locales [1%numLocales] do count = here.runningTasks();
  writeln(count);
}

for param i in 1..10 do movedOn();
```

### CODE REVIEW

File: compiler/optimizations/optimizeOnClauses.cpp [hunk 1]
```diff
@@ -541,8 +542,11 @@ static void addRunningTaskModifiers(void) {
       if (fn->hasFlag(FLAG_FAST_ON) == false) {
         CallExpr* call = findRealOnCall(fn);
         SET_LINENO(call);
-        call->insertBefore(new CallExpr(gChplIncRunningTask));
-        call->insertAfter(new CallExpr(gChplDecRunningTask));
+        if (visited.count(call) == 0) {
```
Comment (mppf) [inline compiler/optimizations/optimizeOnClauses.cpp]: Can we change the call variable above so it has a different name from the loop index variable? (It's unnecessarily confusing, as is, IMO).

Also, where are we doing the increment for non-blocking on statements?

Comment (ronawho) [reply compiler/optimizations/optimizeOnClauses.cpp]: > Can we change the call variable above so it has a different name from the loop index variable? (It's unnecessarily confusing, as is, IMO).
Yeah, updated. 

> Also, where are we doing the increment for non-blocking on statements?

Here, this modifies the running task counter for the outlined function for all non-fast on stmts.

Comment (mppf) [reply compiler/optimizations/optimizeOnClauses.cpp]: Right, thanks. I was reading FLAG_ON_BLOCK and thinking that meant "blocking on" but it just means "on statement { } ".

**Comment by bradcray** [comment]
Does this mean that we can claim task counting is bug free now?  :D :D :D

More seriously, though, were there compelling use cases that were running afoul of this that were mystifying us?

**Comment by ronawho** [comment]
> Does this mean that we can claim task counting is bug free now? :D :D :D

:) I was really hoping you wouldn't notice this one go by. 

> More seriously, though, were there compelling use cases that were running afoul of this that were mystifying us?

Not that I know of. This is a case Michael noticed while looking at the generated code as part getting the unordered compiler optimization to run for zippered/promoted case.

**Comment by mppf** [comment]
> This is a case Michael noticed while looking at the generated code as part getting the unordered compiler optimization to run for zippered/promoted case.

Yeah it just looked "obviously wrong" as there were about 30 increments and decrements in one wrapper function.

**Comment by bradcray** [comment]
OK, thanks for the additional info.

**Pull request merged**